### PR TITLE
[SPARK-21190][PYSPARK] Python Vectorized UDFs 

### PR DESCRIPTION
**Pull request by BryanCutler**
## What changes were proposed in this pull request?

This PR adds vectorized UDFs to the Python API

**Proposed API**
Introduce a flag to turn on vectorization for a defined UDF, for example:

```
@pandas_udf(DoubleType())
def plus(a, b)
    return a + b
```
or

```
plus = pandas_udf(lambda a, b: a + b, DoubleType())
```
Usage is the same as normal UDFs

0-parameter UDFs
pandas_udf functions can declare an optional `**kwargs` and when evaluated, will contain a key "size" that will give the required length of the output.  For example:

```
@pandas_udf(LongType())
def f0(**kwargs):
    return pd.Series(1).repeat(kwargs["size"])

df.select(f0())
```

## How was this patch tested?
Added new unit tests in pyspark.sql that are enabled if pyarrow and Pandas are available.

## TODO
- [x] Fix support for promoted types with null values
- [x] Discuss 0-param UDF API (use of kwargs)
- [x] Add tests for chained UDFs
- [ ] Discuss behavior when pyarrow not installed / enabled
- [ ] Cleanup pydoc and add user docs

### CODE REVIEW

**Comment by BryanCutler** [comment]
The following was used to test performance locally

```python
spark = SparkSession.builder.appName("vectorized_udfs").getOrCreate()

vectorize = True

if vectorize:
    from numpy import log, exp
    spark.conf.set("spark.sql.execution.arrow.vectorizeUDFs", "true")
else:
    from math import log, exp
    spark.conf.set("spark.sql.execution.arrow.vectorizeUDFs", "false")

def my_func(p1, p2):
    w = 0.5
    return exp(log(p1) + log(p2) - log(w))

df = spark.range(1 << 24, numPartitions=16).toDF("id") \
    .withColumn("p1", rand()).withColumn("p2", rand())

my_udf = udf(my_func, DoubleType())

df.withColumn("p", my_udf(col("p1"), col("p2")))
```

**Non-Vectorized** ~ 6.127449s

**Vectorized** ~ 2.867868s

Speedup of 2.14x

**Comment by BryanCutler** [comment]
Some comments on the performance above

1)  I used the `ArrowFileWriter` that is currently in the `ArrowSerializer` - this carries significant overhead in copying data to temporary buffers before transferring.  Using the `ArrowStreamWriter` I was seeing much better performance, however it makes significant changes to PythonRunner in PythonRDD.  If we move forward with this I can present those changes as well.

2)  This naively transfers data back into a GenericInternalRow - I'm sure there is a better way to do this that is probably more efficient, maybe someone more familiar with SQL internals can comment

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala [hunk 1]
```diff
@@ -132,6 +135,61 @@ private[sql] object ArrowConverters {
...
+      loadNextBatch()
+
+      override def hasNext: Boolean = _root != null && _index < _root.getRowCount
+
+      override def next(): InternalRow = {
+        val fields = _root.getFieldVectors.asScala
+
+        val genericRowData = fields.map { field =>
+          field.getAccessor.getObject(_index)
+        }.toArray[Any]
```
Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: How about using `SpecificInternalRow`? I think that it could eliminate some boxing/unboxing. The following is a snippet for this usage.

```java
        val fieldTypes = fields.map { field =>
          field match {
            case NullableIntVector => IntegerType
            case NullableFloat8Vector => DoubleType
            ...
          }
        }
        val row = new SpecificInternalRow(fieldTypes)
        fields.zipWithIndex.map { case (field, i) =>
          field match {
            case NullableIntVector =>
              row.setInt(i, field.asInstanceOf[NullableIntVector].getAccessor.get(_index)) 
            case NullableFloat8Vector => LongType
              row.setDouble(i, field.asInstanceOf[NullableFloat8Vector].getAccessor.get(_index))
            ...  
          }
        }
```

Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: As additional performance optimization, can we reuse `InternalRow` object that `next()` method returns?
For example, I think that this [`next()`](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/WholeStageCodegenExec.scala#L399) reuses `UnsafeRow` that is allocated in Java code, which is generated at [here ](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/GenerateUnsafeProjection.scala#L319), as an instance variable.

Comment (BryanCutler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: Thanks @kiszk , I'll give that a shot and see if it helps!

Comment (icexelloss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: @BryanCutler,

I have implemented arrow -> unsafe row conversions in:

https://github.com/icexelloss/spark/commit/8f38c15ea372c1ec9c2fefc36c0bfc3a22c3be14#diff-52cca47e7a940849b28d476ddf99d65eR575

This reuses the row object and doesn't do boxing. Hopefully it's useful to you as well?

Comment (kiszk) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: @BryanCutler
As @cloud-fan suggested [here](https://issues.apache.org/jira/browse/SPARK-21190?focusedCommentId=16101014), it is good to create `ColumnarBatch` with `ArrowColumnVector` and get an iterator like this:
cc: @ueshin

```
    new Iterator[InternalRow] {
      private val _allocator = new RootAllocator(Long.MaxValue)
      private var _reader: ArrowFileReader = _
      private var _root: VectorSchemaRoot = _
      private var _index = 0
      private var _iterator = null

      loadNextBatch()

      override def hasNext: Boolean = _root != null && _index < _root.getRowCount && _iterator.hasNext

      override def next(): InternalRow = {
        _index += 1
        if (_index >= _root.getRowCount) {
          _index = 0
          loadNextBatch()
          if (!hasNext) {
            close()
          }
        }
        _iterator.next()
      }
      ...
      private def loadNextBatch(): Unit = {
        closeReader()
        if (iter.hasNext) {
          val in = new ByteArrayReadableSeekableByteChannel(iter.next().asPythonSerializable)
          _reader = new ArrowFileReader(in, _allocator)
          _root = _reader.getVectorSchemaRoot // throws IOException
          _reader.loadNextBatch() // throws IOException
          val arrowSchema = ArrowUtils.fromArrowSchema(_root.getSchema)
          val fields = _root.getFieldVectors
          val rows = _root.getRowCount
          val columnarBatch = ColumnarBatch.allocateArrow(
            _root.getFieldVectors.asInstanceOf[java.util.List[ValueVector]],
            ArrowUtils.fromArrowSchema(_root.getSchema), _root.getRowCount)
          _iterator = columnarBatch.rowIterator
        }
      }

public final class ColumnarBatch {
...
  public static ColumnarBatch allocateArrow(List<ValueVector> vectors, StructType schema, int maxRows) {
    // need to implement the following constructor for arrowColumnVector
    return new ColumnarBatch(vectors, schema, maxRows);
  }
...
}
```

Comment (BryanCutler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: Thanks @kiszk , I'm giving it a try!

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java [hunk 2]
```diff
@@ -65,15 +65,44 @@
...
+  private static ColumnVector[] allocateVectors(StructType schema, int maxRows, MemoryMode memMode) {
+    ColumnVector[] columns = new ColumnVector[schema.size()];
+    for (int i = 0; i < schema.fields().length; ++i) {
+      StructField field = schema.fields()[i];
+      columns[i] = ColumnVector.allocate(maxRows, field.dataType(), memMode);
+    }
+    return columns;
+  }
+
+  public static ColumnarBatch createReadOnly(
```
Comment (BryanCutler) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: @ueshin I made some changes here to allow for use with `ArrowColumnVectors`.  I was thinking of putting these in a separate JIRA because it can be used regardless of what is done with vectorized UDFs.  What do you think?

Comment (ueshin) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: @BryanCutler I agree with you, let's separate it from this pr.

Comment (BryanCutler) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java]: ok, will do.  I created https://issues.apache.org/jira/browse/SPARK-21583 for this

File: python/pyspark/sql/functions.py [hunk 3]
```diff
@@ -2112,7 +2113,7 @@ def wrapper(*args):
 
 
 @since(1.3)
-def udf(f=None, returnType=StringType()):
+def udf(f=None, returnType=StringType(), vectorized=False):
```
Comment (BryanCutler) [inline python/pyspark/sql/functions.py]: @felixcheung does this fit your idea for a more generic decorator?  Not exclusively labeled as `pandas_udf`, just enable vectorization with a flag, e.g. `@udf(DoubleType(), vectorized=True)`

**Comment by felixcheung** [comment]
Cool!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+          pyFuncs, bufferSize, reuseWorker, PythonEvalType.SQL_ARROW_UDF, argOffsets).
+        compute(inputIterator, context.partitionId(), context)
+
+      val joined = new JoinedRow
+      val resultProj = UnsafeProjection.create(output, output)
+
+      val outputRowIterator = ArrowConverters.fromPayloadIterator(
+        outputIterator.map(new ArrowPayload(_)), context)
+
+      assert(schemaOut.equals(outputRowIterator.schema))
```
Comment (BryanCutler) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: @felixcheung , I think you had also brought up checking the return type matches what was defined in the UDF.  This is done here.

File: core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala [hunk 5]
```diff
@@ -83,10 +83,23 @@ private[spark] case class PythonFunction(
  */
 private[spark] case class ChainedPythonFunctions(funcs: Seq[PythonFunction])
 
+/**
+ * Enumerate the type of command that will be sent to the Python worker
+ */
+private[spark] object PythonEvalType {
+  val NON_UDF = 0
+  val SQL_BATCHED_UDF = 1
+  val SQL_ARROW_UDF = 2
```
Comment (cloud-fan) [inline core/src/main/scala/org/apache/spark/api/python/PythonRDD.scala]: the new udf parameter is pandas `Series`, I think it's more accurate to call it `SQL_PANDAS_UDF`.

File: python/pyspark/serializers.py [hunk 6]
```diff
@@ -81,6 +81,12 @@ class SpecialLengths(object):
     NULL = -5
 
 
+class PythonEvalType(object):
+    NON_UDF = 0
+    SQL_BATCHED_UDF = 1
+    SQL_ARROW_UDF = 2
```
Comment (cloud-fan) [inline python/pyspark/serializers.py]: ditto

File: python/pyspark/sql/functions.py (hunk 3 above)
Comment (cloud-fan) [reply python/pyspark/sql/functions.py]: I think `@pandas_udf(DoubleType())` is better than `@udf(DoubleType(), vectorized=True)`, which is more concise.

**Comment by cloud-fan** [comment]
I prefer this PR over #19147 because it's much smaller, but we do need some tests. @ueshin can you contribute your test cases to this PR? thanks!

File: python/pyspark/sql/functions.py (hunk 3 above)
Comment (cloud-fan) [reply python/pyspark/sql/functions.py]: as we discussed in the email, we should also accept data type of string format.

Comment (cloud-fan) [reply python/pyspark/sql/functions.py]: and also `**kwargs` to bring the size information

**Comment by HyukjinKwon** [comment]
(I am sorry, I didn't realise this PR was open already ..)

File: python/pyspark/sql/functions.py (hunk 3 above)
Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: It seems like the consensus is for `pandas_udf` and I'm fine with that too.  I'll make that change and the others brought up here.

**Comment by BryanCutler** [comment]
Thanks @cloud-fan.  @ueshin , it would be great to incorporate the tests you have, let me know what is easiest for you.  I'll push an update to change to `@pandas_udf` so hopefully the tests would run off the bat, but if not I can help sort it out.

**Comment by ueshin** [comment]
@BryanCutler I sent a pr to your repository https://github.com/BryanCutler/spark/pull/26. Could you take a look at it please?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala [hunk 7]
```diff
@@ -82,7 +82,6 @@ private[sql] object ArrowConverters {
 
     val root = VectorSchemaRoot.create(arrowSchema, allocator)
     val arrowWriter = ArrowWriter.create(root)
-
```
Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/arrow/ArrowConverters.scala]: (Looks unrelated change)

File: python/pyspark/serializers.py [hunk 8]
```diff
@@ -199,6 +211,33 @@ def __repr__(self):
...
+class ArrowPandasSerializer(ArrowSerializer):
+
+    def __init__(self):
+        super(ArrowPandasSerializer, self).__init__()
+
+    def dumps(self, series):
+        """
+        Make an ArrowRecordBatch from a Pandas Series and serialize
+        """
+        import pyarrow as pa
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py]: Should we catch `ImportError`?

**Comment by BryanCutler** [comment]
@ueshin , I merged your tests and added support for `**kwargs` to use "size" for 0-parameter UDFs.  Do you think this might be a little better to be called "length" or "output_length"?

I still need to fix the tests with null values, it shouldn't be a problem with the fix you provided.  I'll add that soon. Otherwise I believe other tests are working, but we probably need to check chained UDFs also.

File: python/pyspark/serializers.py (hunk 8 above)
Comment (BryanCutler) [reply python/pyspark/serializers.py]: Yeah, it would probably be best to handle it the same way as in `toPandas()`.

That got me thinking that it is a little weird to have an SQLConf "spark.sql.execution.arrow.enable" that is set for `toPandas()` but has no bearing with `pandas_udf`.  It doesn't need to since it is an explicit call but seems a little contradictory, what do you think?

Comment (HyukjinKwon) [reply python/pyspark/serializers.py]: Ah, I see. Thanks for explaining it. Sure, I am okay with leaving it as is here.
I think we should catch and throw it with better messages in all cases (probably in the entry points) but probably let's talk about this later in another place.

Comment (HyukjinKwon) [reply python/pyspark/serializers.py]: Ah, hm .. let me check the previous discussions and think about this a bit more.

Comment (HyukjinKwon) [reply python/pyspark/serializers.py]: I am okay with leaving it as is here. I think we should catch and throw it with better messages in all cases later but let's talk about this in another place.

**Comment by ueshin** [comment]
@BryanCutler I think it's okay to rename `size` to `length` (or longer name to avoid name-conflict like `_length_`?).

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+ * A physical plan that evaluates a [[PythonUDF]],
+ */
+case class ArrowEvalPythonExec(udfs: Seq[PythonUDF], output: Seq[Attribute], child: SparkPlan)
+  extends SparkPlan {
+
+  def children: Seq[SparkPlan] = child :: Nil
+
+  override def producedAttributes: AttributeSet = AttributeSet(output.drop(child.output.length))
+
+  private def collectFunctions(udf: PythonUDF): (ChainedPythonFunctions, Seq[Expression]) = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: `producedAttributes` and `collectFunctions` looks duplicate between `ArrowEvalPythonExec` and `BatchEvalPythonExec`. We can duplicate them, maybe in later PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+          }
+        }.toArray
+      }.toArray
+      val projection = newMutableProjection(allInputs, child.output)
+      val schema = StructType(dataTypes.zipWithIndex.map { case (dt, i) =>
+        StructField(s"_$i", dt)
+      })
+
+      // Input iterator to Python: input rows are grouped so we send them in batches to Python.
+      // For each row, add it to the queue.
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: The comment is wrong now. We don't group input rows here.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+          projectedRowIter, schema, conf.arrowMaxRecordsPerBatch, context).
+        map(_.asPythonSerializable)
+
+      val schemaOut = StructType.fromAttributes(output.drop(child.output.length).zipWithIndex.
+        map { case (attr, i) => attr.withName(s"_$i") })
+
+      // Output iterator for results from Python.
+      val outputIterator = new PythonRunner(
+          pyFuncs, bufferSize, reuseWorker, PythonEvalType.SQL_PANDAS_UDF, argOffsets).
+        compute(inputIterator, context.partitionId(), context)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: nit: I think we usually write it in a style as:

    val outputIterator = new PythonRunner(
        pyFuncs, bufferSize, reuseWorker, PythonEvalType.SQL_PANDAS_UDF, argOffsets)
      .compute(inputIterator, context.partitionId(), context)

There are similar styles above, e.g. `map { case (attr, i) => ...` -> `.map { case (attr, i) => ...`.

File: python/pyspark/sql/functions.py [hunk 12]
```diff
@@ -2142,18 +2159,26 @@ def udf(f=None, returnType=StringType()):
...
+
+    :param f: python function if used as a standalone function
+    :param returnType: a :class:`pyspark.sql.types.DataType` object
+
+    # TODO: doctest
+    """
+    import inspect
+    # If function "f" does not define the optional kwargs, then wrap with a kwargs placeholder
+    if inspect.getargspec(f).keywords is None:
+        return _create_udf(lambda *a, **kwargs: f(*a), returnType=returnType, vectorized=True)
```
Comment (viirya) [inline python/pyspark/sql/functions.py]: We wrap a kwargs placeholder for the function, but we don't actually pass it into the function. So different than the 0-argument pandas udf in SPIP, we explicitly ask it to define a kwargs? Namely we don't have really 0-argument pandas udf, because it at least has kwargs defined?

Comment (viirya) [reply python/pyspark/sql/functions.py]: Should we make the kwargs as mandatory for 0-argument pandas udf? I think a 0-argument pandas udf without the kwargs seems no making sense as it can't guess the size of returning Series.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala (hunk 9 above)
Comment (BryanCutler) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/ArrowEvalPythonExec.scala]: Yes, these functions are duplicated, as well as some code in `doExecute()`.  I could add a common base class like `EvalPythonExec` to clean this up, and maybe move to the same file?

File: python/pyspark/sql/functions.py (hunk 12 above)
Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: If the user function doesn't define the keyword args, then it is wrapped with a placeholder so that `worker.py` can expect the function to always have keywords.  I thought this was better than trying to do inspection on the worker while running the UDF.

I'm not crazy about the 0-parameter pandas_udf, but if we have to support it here then it does need to get the required length of output somehow, unless we repeat/slice the output to make the length correct.

I'm ok with making `**kwargs` mandatory for 0-parameter UDFs and optional for others.

File: python/pyspark/sql/tests.py [hunk 13]
```diff
@@ -3122,6 +3122,185 @@ def test_filtered_frame(self):
...
+        schema = StructType().add("str", StringType())
+        df = self.spark.createDataFrame(data, schema)
+        str_f = pandas_udf(lambda x: x, StringType())
+        res = df.select(str_f(col('str')))
+        self.assertEquals(df.collect(), res.collect())
+
+    def test_vectorized_udf_zero_parameter(self):
+        from pyspark.sql.functions import pandas_udf
+        import pandas as pd
+        df = self.spark.range(100000)
```
Comment (BryanCutler) [inline python/pyspark/sql/tests.py]: @ueshin, was there any particular reason for such a high number or would it be fine if I reduced the range to 10?

File: python/pyspark/serializers.py [hunk 14]
```diff
@@ -199,6 +211,46 @@ def __repr__(self):
...
+        Make an ArrowRecordBatch from a Pandas Series and serialize. Input is a single series or
+        a list of series accompanied by an optional pyarrow type to coerce the data to.
+        """
+        import pyarrow as pa
+        # Make input conform to [(series1, type1), (series2, type2), ...]
+        if not isinstance(series, (list, tuple)) or \
+                (len(series) == 2 and isinstance(series[1], pa.DataType)):
+            series = [series]
+        series = [(s, None) if not isinstance(s, (list, tuple)) else s for s in series]
+        arrs = [pa.Array.from_pandas(s[0], type=s[1], mask=s[0].isnull()) for s in series]
```
Comment (BryanCutler) [inline python/pyspark/serializers.py]: With pyarrow 0.7, it is now able to internally cast the type if needed, so this is working.

**Comment by BryanCutler** [comment]
@ueshin , the tests are all passing now when using pyarrow 0.7 (just released).  This added better support for type coercion in `Array.from_pandas` which makes handling null values a little easier.  What are your thoughts on upgrading?  We will probably want to do so to get some improvements for Decimal types anyway.

File: python/pyspark/sql/tests.py (hunk 13 above)
Comment (ueshin) [reply python/pyspark/sql/tests.py]: I think it's fine to reduce the range.

**Comment by ueshin** [comment]
@BryanCutler I'm ok to upgrade pyarrow to 0.7 except for the same concerns as #18974.
I guess we need to discuss upgrade policy and strategy of pyarrow.

File: python/pyspark/serializers.py [hunk 15]
```diff
@@ -199,6 +211,46 @@ def __repr__(self):
...
+        batch = pa.RecordBatch.from_arrays(arrs, ["_%d" % i for i in range(len(arrs))])
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py]: I'd use `xrange`.

File: python/pyspark/serializers.py [hunk 16]
```diff
@@ -199,6 +211,46 @@ def __repr__(self):
...
+        """
+        Make an ArrowRecordBatch from a Pandas Series and serialize. Input is a single series or
+        a list of series accompanied by an optional pyarrow type to coerce the data to.
+        """
+        import pyarrow as pa
+        # Make input conform to [(series1, type1), (series2, type2), ...]
+        if not isinstance(series, (list, tuple)) or \
+                (len(series) == 2 and isinstance(series[1], pa.DataType)):
+            series = [series]
+        series = [(s, None) if not isinstance(s, (list, tuple)) else s for s in series]
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py]: I'd use generator comprehension.

File: python/pyspark/serializers.py [hunk 17]
```diff
@@ -199,6 +211,46 @@ def __repr__(self):
...
+    def loads(self, obj):
+        """
+        Deserialize an ArrowRecordBatch to an Arrow table and return as a list of pandas.Series
+        followed by a dictionary containing length of the loaded batches.
+        """
+        import pyarrow as pa
+        reader = pa.RecordBatchFileReader(pa.BufferReader(obj))
+        batches = [reader.get_batch(i) for i in range(reader.num_record_batches)]
+        # NOTE: a 0-parameter pandas_udf will produce an empty batch that can have num_rows set
+        num_rows = sum([batch.num_rows for batch in batches])
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py]: I'd use generator comprehension here too.

File: python/pyspark/serializers.py [hunk 18]
```diff
@@ -199,6 +211,46 @@ def __repr__(self):
...
+        return super(ArrowPandasSerializer, self).dumps(batch)
+
+    def loads(self, obj):
+        """
+        Deserialize an ArrowRecordBatch to an Arrow table and return as a list of pandas.Series
+        followed by a dictionary containing length of the loaded batches.
+        """
+        import pyarrow as pa
+        reader = pa.RecordBatchFileReader(pa.BufferReader(obj))
+        batches = [reader.get_batch(i) for i in range(reader.num_record_batches)]
```
Comment (HyukjinKwon) [inline python/pyspark/serializers.py]: And .. `xrange` here too

File: python/pyspark/sql/functions.py (hunk 12 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py]: How about disallowing it for now? I think it could be an option if 0-parameter UDF alone should not be supported consistently. `return pd.Series(1).repeat(kwargs['length'])` looks still a little bit weird ..

**Comment by cloud-fan** [comment]
what if users installed an older version of pyarrow? Shall we throw exception and ask them to upgrade, or work around type casting issue?

File: python/pyspark/worker.py [hunk 19]
```diff
@@ -71,7 +73,19 @@ def wrap_udf(f, return_type):
...
-def read_single_udf(pickleSer, infile):
+def wrap_pandas_udf(f, return_type):
+    def verify_result_length(*a):
+        kwargs = a[-1]
+        result = f(*a[:-1], **kwargs)
+        if len(result) != kwargs["length"]:
+            raise RuntimeError("Result vector from pandas_udf was not the required length: "
+                               "expected %d, got %d\nUse input vector length or kwarg['length']"
+                               % (kwargs["length"], len(result)))
+        return result, toArrowType(return_type)
```
[the remark this answers is not in the document]
Comment (BryanCutler) [reply python/pyspark/worker.py]: sure, that sounds good thanks!

File: python/pyspark/serializers.py (hunk 16 above)
Comment (BryanCutler) [reply python/pyspark/serializers.py]: That would work, but does it help much since `series` will already be a list or tuple?

File: python/pyspark/serializers.py (hunk 17 above)
Comment (BryanCutler) [reply python/pyspark/serializers.py]: I guess this makes sense because its a summation, no sense in making a list then adding it all up

**Comment by BryanCutler** [comment]
Thanks for the reviews @ueshin @viirya and @HyukjinKwon !  I updated with your comments

File: python/pyspark/sql/functions.py (hunk 12 above)
Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: I agree it is still a bit weird.. Did you mean disallowing 0-parameter panda_udfs or requiring 0-parameter panda_udfs to accept `kwargs`?

**Comment by BryanCutler** [comment]
Regarding the upgrade of Arrow, the concerns of #18974 are still valid - namely it has some risk and upgrading the Python side is a good amount of work that only a couple of people have the access to do.  Would it be better to discuss the upgrade strategy in another JIRA?

**Comment by BryanCutler** [comment]
> what if users installed an older version of pyarrow? Shall we throw exception and ask them to upgrade, or work around type casting issue?

@cloud-fan , in regards to handling of problems that might come up if using different versions of Arrow, I think we should first decide on a minimum supported version, then maybe we could put that version of pyarrow as a requirement for PySpark.  If we decide to use 0.4.1 which we currently use, then we should probably work around the type casting issue and make sure this PR works with that version.

File: python/pyspark/serializers.py (hunk 16 above)
Comment (HyukjinKwon) [reply python/pyspark/serializers.py]: Yea, it actually affects the performance because we can avoid an extra loop:

```python
def im_map(x):
    print("I am map %s" % x)
    return x

def im_gen(x):
    print("I am gen %s" % x)
    return x

def im_list(x):
    print("I am list %s" % x)
    return x

items = list(range(3))
map(im_map, [im_list(item) for item in items])
map(im_map, (im_gen(item) for item in items))
```

And .. this actually affects the performance up to my knowledge:

```python
import time

items = list(xrange(int(1e8)))

for _ in xrange(10):
    s = time.time()
    _ = map(lambda x: x, [item for item in items])
    print "I am list comprehension with a list: %s" % (time.time() - s)
    s = time.time()
    _ = map(lambda x: x, (item for item in items))
    print "I am generator expression with a list: %s" % (time.time() - s)
```

This gives me ~13% improvement :).

Comment (HyukjinKwon) [reply python/pyspark/serializers.py]: This might not be a big deal but .. I usually use generator if it iterates once and is discarded. This should consume less memory too as list comprehension should be evaluated once first.

File: python/pyspark/sql/functions.py (hunk 12 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/functions.py]: Ah, I was thinking that disallowing 0-parameter panda_udf could be an option ...

**Comment by cloud-fan** [comment]
ok let's work around the type casting issue and discuss arrow upgrading later.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+ *
+ *            Downstream (for parent)
+ *             /      \
+ *            /     socket  (output of UDF)
+ *           /         \
+ *        RowQueue    Python
+ *           \         /
+ *            \     socket  (input of UDF)
+ *             \     /
+ *          upstream (from child)
```
Comment (LiShuMing) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Is `Upstream` better?

Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: I think `upstream` is fine.

Comment (LiShuMing) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: Maybe I put myself uncomfortable to see `Downstream` upper, forgive me..

File: python/pyspark/serializers.py (hunk 16 above)
Comment (BryanCutler) [reply python/pyspark/serializers.py]: Thanks @HyukjinKwon , I suppose if there are more than a few series then it might make some difference.  In that case, every little bit helps so sounds good to me!

File: python/pyspark/sql/functions.py (hunk 12 above)
Comment (BryanCutler) [reply python/pyspark/sql/functions.py]: It would be a lot cleaner to just not allow 0-parameters. Is it an option to not allow 0-parameter UDFs for pandas_udfs @ueshin @cloud-fan ?

**Comment by BryanCutler** [comment]
@ueshin I haven't had much luck with the casting workaround:
```
pa.Array.from_pandas(s.astype(t.to_pandas_dtype(), copy=False), mask=s.isnull(), type=t)
```
It appears that it forces a copy for floating point -> integer and then checks if any NaNs, so I get the error `ValueError: Cannot convert non-finite values (NA or inf) to integer`.  I'm using Pandas 0.20.1, but also tried 0.19.4 with the same result, any ideas?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala (hunk 20 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/python/EvalPythonExec.scala]: that's fine but either looks fine and not a big deal.

**Comment by ueshin** [comment]
@BryanCutler Hmm, I'm not exactly sure the reason why it doesn't work (or mine works) but we can use `fillna(0)` before casting like:

```
pa.Array.from_pandas(s.fillna(0).astype(t.to_pandas_dtype(), copy=False), mask=s.isnull(), type=t)
```

**Comment by BryanCutler** [comment]
Thanks @ueshin , that works to allow the tests to pass. I do worry that it might cause some other issues and I would much prefer we upgrade Arrow to handle this, but I'll push this and we can discuss.

File: python/pyspark/sql/functions.py (hunk 12 above)
Comment (cloud-fan) [reply python/pyspark/sql/functions.py]: I'm fine to disallow 0-parameter pandas udf, as it's not a common case. We can add it when people request it.

File: python/pyspark/serializers.py [hunk 21]
```diff
@@ -199,6 +211,55 @@ def __repr__(self):
         return "ArrowSerializer"
 
 
+class ArrowPandasSerializer(ArrowSerializer):
+    """
+    Serializes Pandas.Series as Arrow data.
+    """
+
+    def __init__(self):
+        super(ArrowPandasSerializer, self).__init__()
```
Comment (ueshin) [inline python/pyspark/serializers.py]: Do we need this?

File: python/pyspark/serializers.py [hunk 22]
```diff
@@ -199,6 +211,55 @@ def __repr__(self):
...
+        """
+        Make an ArrowRecordBatch from a Pandas Series and serialize. Input is a single series or
+        a list of series accompanied by an optional pyarrow type to coerce the data to.
+        """
+        import pyarrow as pa
+        # Make input conform to [(series1, type1), (series2, type2), ...]
+        if not isinstance(series, (list, tuple)) or \
+                (len(series) == 2 and isinstance(series[1], pa.DataType)):
+            series = [series]
+        series = ((s, None) if not isinstance(s, (list, tuple)) else s for s in series)
```
Comment (cloud-fan) [inline python/pyspark/serializers.py]: `s for s in series` is same as `series`, isn't it?

**Comment by cloud-fan** [comment]
LGTM, merging to master!

We can address remaining minor comments in follow-up, and have new PRs to remove the 0-parameter UDF and use arrow streaming protocol.

**Comment by BryanCutler** [comment]
Thanks @cloud-fan @ueshin and others who reviewed!  I'll make followups to disable 0-param and complete the docs for this.

File: python/pyspark/serializers.py (hunk 21 above)
Comment (BryanCutler) [reply python/pyspark/serializers.py]: No, that was leftovers.. I'll remove it in a followup.
